rust_src/src/keymap.rs

```diff
@@ -317,6 +316,92 @@ pub fn map_keymap_lisp(function: LispObject, keymap: LispObject, sort_first: boo
     LispObject::constant_nil()
 }
 
+/// Call FUN for every binding in MAP and stop at (and return) the parent.
+/// FUN is called with 4 arguments: FUN (KEY, BINDING, ARGS, DATA).
+#[no_mangle]
+pub extern "C" fn map_keymap_internal(
+    map: Lisp_Object,
+    fun: map_keymap_function_t,
+    args: Lisp_Object,
+    data: *const c_void,
+) -> Lisp_Object {
+    let map = LispObject::from_raw(map);
+    let mut tail = match map.as_cons() {
+        None => LispObject::constant_nil(),
+        Some(cons) => {
+            let (car, cdr) = cons.as_tuple();
+            if car.eq_raw(Qkeymap) {
+                cdr
+            } else {
+                map
+            }
+        }
+    };
+
+    for tail_cons in tail.iter_tails_safe() {
+        let binding = tail_cons.car();
+        if binding.eq_raw(Qkeymap) {
+            break;
+        } else {
+            tail = tail_cons.cdr();
```

Comment:
ok, i see my confusion. This _looks_ like you are trying to affect the loop. Which can't happen because you are using an iterator. But what you are really trying to do is update tail for the eventual return. The right thing to do here is use a different variable name for the return value.

Reply:
Makes sense. Alright ;)